extensions/firmware/org.eclipse.smarthome.firmware.filesystem/src/main/java/org/eclipse/smarthome/firmware/filesystem/FileSystemFirmwareProvider.java

```diff
@@ -0,0 +1,291 @@
+                            FirmwareUID firmwareUID = new FirmwareUID(thingTypeUID, version);
+                            firmwareFiles.put(firmwareUID, getBaseName(file.getFileName().toString()));
+                        } else {
+                            logger.trace("File {} does not contain a version entry", file.getFileName().toString());
+                        }
+                    }
+                    return FileVisitResult.CONTINUE;
+                }
+            });
+        } catch (IOException e) {
+            logger.warn("Error reading directory {}", rootPath);
+        }
+    }
+
+    @Override
+    public Firmware getFirmware(FirmwareUID firmwareUID) {
+        return getFirmware(firmwareUID, Locale.getDefault());
+    }
+
+    @Override
+    public Firmware getFirmware(FirmwareUID firmwareUID, Locale locale) {
+        Preconditions.checkArgument(firmwareUID != null);
+        if (locale == null) {
+            locale = Locale.getDefault();
+        }
+
+        String filename = firmwareFiles.get(firmwareUID);
+        if (filename == null) {
+            logger.debug("Cannot find firmware {}", firmwareUID);
+            return null;
+        }
+
+        try {
+            return getFirmware(firmwareUID, filename, locale);
+        } catch (IOException e) {
+            logger.error("Cannot find firmware " + firmwareUID, e);
+        }
+
+        return null;
+    }
+
+    private Firmware getFirmware(FirmwareUID firmwareUID, String filename, Locale locale)
+            throws IOException {
+        ThingTypeUID thingTypeUID = firmwareUID.getThingTypeUID();
+        URL[] urls = { Paths.get(rootPath, thingTypeUID.getBindingId(), thingTypeUID.getId()).toUri().toURL() };
+        ResourceBundle rb = ResourceBundle.getBundle(filename, locale, new URLClassLoader(urls));
+
+        InputStream firmwareImage = getFirmwareImage(firmwareUID, rb.getString("image"));
+
+        return new Firmware.Builder(firmwareUID).withDescription(getResourceAsString(rb, "description"))
+                .withModel(getResourceAsString(rb, "model"))
+                .withVendor(getResourceAsString(rb, "vendor")).withChangelog(getResourceAsString(rb, "changelog"))
+                .withPrerequisiteVersion(getResourceAsString(rb, "prerequisiteVersion"))
+                .withOnlineChangelog(new URL(getResourceAsString(rb, "onlineChangelog"))).withContent(firmwareImage)
+                .build();
+    }
+
+    private InputStream getFirmwareImage(FirmwareUID firmwareUID, String filename) throws IOException {
+        ThingTypeUID thingTypeUID = firmwareUID.getThingTypeUID();
+        Path imageFile = Paths.get(rootPath, thingTypeUID.getBindingId(), thingTypeUID.getId(), filename);
+    
+        if (Files.notExists(imageFile)) {
+            throw new IOException("File " + imageFile.getFileName().toString() + " cannot be found");
+        }
+    
+        return FileUtils.openInputStream(imageFile.toFile());
```

Comment:
Here the imageFile can be checked if a given value matches the MD5 or SHA1 hash value